src/sentry/event_manager.py

```diff
@@ -295,15 +320,90 @@ class InvalidTimestamp(Exception):
 
 
 class EventManager(object):
-    logger = logging.getLogger('sentry.events')
+    """
+    Handles normalization in both the store endpoint and the save task. The
+    intention is to swap this class out with a reimplementation in Rust.
+    """
 
     def __init__(self, data, version='5'):
-        self.data = CanonicalKeyDict(data)
+        if not isinstance(data, (bytes, six.text_type)):
+            data = CanonicalKeyDict(data)
+        self._data = data
```

Comment:
_nit_: Just store the raw data here (e.g. in `_raw`), since `decode` needs to be called anyway.

Reply:
The silly thing is that EventManager is also initialized with a JSON hash or some mixture with interfaces (in tasks/store). The idea is that this is represented as `enum { RawPayload(..), JsonValue(..), Structured(..) }` in Rust.

Not sure whether we want to split this up in two classes maybe

Reply:
I see. Splitting it up might actually make sense now. We're not going to migrate `save` to Rust, are we?

Reply:
I'm not entirely sure if we have a good reason for splitting this up now though. It would require readjustments to a lot of APIs and there are simple ways around this inconvenience for our purposes.

Reply:
Partially fixed: decode() happens in init, so `get_data()` no longer returns different types